### PR TITLE
Issue 17: Stopping at EOS

### PR DESCRIPTION
**Pull request by aKssup**
- Updated `run_pm.sh` to include the correct environment name. Initially the environment is named "yalis_venv" when running the `create_python_env_perlmutter.sh` script, hence "yalis_venv_v2" doesn't work.
- Adding single batch and multi-batch logic. For a single batch, the generation loop short circuits upon detecting EOS. For multiple batches, with the use of tensor masks, we wait until all batches to hit EOS for stopping the generation loop. 
- Tested the code with Meta-Llama-3-8B-Instruct and it runs.

Follow-ups:
- I create a temporary PyTorch tensor that holds the eos tokens list (depending on the model, for now we specify it to be the meta llama 3 8B in `infer.py`). When initializing the tensor, I set the device to "cuda". Is that a valid approach? The tensor is essentially used to create a mask that stores information on whether EOS is detected or not.

### CODE REVIEW

**Comment by siddharth9820** [comment]
Thanks! I'll review this tomorrow. Can you please post the output tok/s vs batch size?

Different batch sizes can be used by using different number of prompts from infer.py

**Comment by aKssup** [comment]
> Thanks! I'll review this tomorrow. Can you please post the output tok/s vs batch size?
> 
> Different batch sizes can be used by using different number of prompts from infer.py

tok:s vs batch size yalis issue 17

Averaged values:
Batch size = 1: 155.10 tok/sec
Batch size = 2: 274.25 tok/sec
Batch size = 4: 526.41 tok/sec
Batch size = 8: 990.22 tok/sec

**Comment by aKssup** [comment]
> > Thanks! I'll review this tomorrow. Can you please post the output tok/s vs batch size?
> > Different batch sizes can be used by using different number of prompts from infer.py
> 
> tok:s vs batch size yalis issue 17
> 
> Averaged values: Batch size = 1: 155.10 tok/sec Batch size = 2: 274.25 tok/sec Batch size = 4: 526.41 tok/sec Batch size = 8: 990.22 tok/sec

image

Here are the average tok/s values:
EOS Detection On:
Batch size 1: 138.10 tok/s
Batch size 2: 231.28 tok/s
Batch size 4: 411.53 tok/s
Batch size 8: 583.04 tok/s

EOS Detection Off:
Batch size 1: 140.24 tok/s
Batch size 2: 234.55 tok/s
Batch size 4: 419.84 tok/s
Batch size 8: 588.17 tok/s

File: yalis/engine.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def __init__(
         self,
         model_config: ModelConfig,
         inference_config: InferenceConfig,
+        generation_config,
```
Comment (siddharth9820) [inline yalis/engine.py]: Let's not make the user pass the generation_config in engine. model_config already has the name of the model in one of its attributes. We can use it to initialize the generation_config inside the engine.

File: yalis/engine.py [hunk 2]
```diff
@@ -253,6 +262,25 @@ def generate(
                         next_token
                     )  # Copy the new token into tokens
                 output_tokens.append(next_token.clone())
+
+                eos_tokens_tensor = torch.tensor(self.generation_config.eos_token_id, device="cuda")
+                # Check ignore_eos tag
+                if not ignore_eos:
+                    # Using the view function to faltten the array because next_token is a 2D tensor.
```
Comment (siddharth9820) [inline yalis/engine.py]: spell mistake

File: yalis/engine.py [hunk 3]
```diff
@@ -253,6 +262,25 @@ def generate(
...
+                    next_token_flatten = next_token.view(-1)
+                    temp_mask = torch.isin(next_token_flatten, eos_tokens_tensor)
+                    # Single-batch stop
+                    if batch_size == 1:
```
Comment (siddharth9820) [inline yalis/engine.py]: Let's only enable eos_detection for batch size 1. If the user sets ignore_eos to False with batch size >1 , let's throw a not supported error.

File: yalis/config.py [hunk 4]
```diff
@@ -89,6 +89,7 @@ def __init__(
         top_p: Optional[float] = 1.0,
         temperature: Optional[float] = 1.0,
         metrics: bool = False,
+        ignore_eos: Optional[bool] = False
```
Comment (siddharth9820) [inline yalis/config.py]: let's make the default True

**Review by siddharth9820 — CHANGES_REQUESTED** [review]
LGTM

File: yalis/engine.py [hunk 5]
```diff
@@ -264,6 +265,14 @@ def generate(
             f"Tokenization took {timers.get_times()[0][('tokenize',)]} ms"
         )
 
+        # Using tensor size instead of inference config object to allow users to choose batch sizes < max batch size
+        # (later defined in inference config).
+        batch_size = prompt_tokens.size(0)
+        ignore_eos = self.inference_config.ignore_eos
```
Comment (prajwal1210) [inline yalis/engine.py]: Why does ignore_eos need to be a config variable? I think just passing it to generate will do. Also allows the flexibility to use the same initialized model with eos/non-eos generation

**Review by prajwal1210 — CHANGES_REQUESTED** [review]

File: yalis/engine.py [hunk 6]
```diff
@@ -180,6 +180,7 @@ def _initialize_model(self):
             print_rank0(
                 "Pad token not found in the tokenizer. Using eos_token as pad token."
             )
+        self.eos_token_id = self.tokenizer.eos_token_id
```
Comment (prajwal1210) [inline yalis/engine.py]: This needs to be removed

File: yalis/engine.py [hunk 7]
```diff
@@ -199,6 +200,7 @@ def generate(
         prompts: Union[list[str], list[list[int]]],
         tokens_to_generate: int = 50,
         report_throughput: bool = False,
+        ignore_eos_param: Optional[bool] = True
```
Comment (prajwal1210) [inline yalis/engine.py]: Just name it "ignore_eos"

File: yalis/engine.py [hunk 8]
```diff
@@ -264,6 +266,14 @@ def generate(
...
         )
 
+        # Using tensor size instead of inference config object to allow users to choose batch sizes < max batch size
+        # (later defined in inference config).
+        batch_size = prompt_tokens.size(0)
+        ignore_eos = ignore_eos_param
+        # Initialize done mask for multiple batches
+        if batch_size > 1 and not ignore_eos:
+            done_mask = torch.zeros(batch_size, dtype=torch.bool, device=self.device)
+        finished_reason = "Max Token Length"
```
Comment (prajwal1210) [inline yalis/engine.py:275]: Where are you updating this to reflect that the finish reason is EOS? I can only see it being set to a value here

File: yalis/engine.py [hunk 9]
```diff
@@ -306,7 +315,23 @@ def generate(
                         next_token
                     )  # Copy the new token into tokens
 
+                # EOS Support:
+                # Flatten to shape (batch_size,) for element wise comparison
+                done_mask |= (next_token.view(-1) == self.tokenizer.eos_token_id)
+                # Reshape to match next_token's shape for masked_fill()
+                mask = done_mask.view(-1, 1)
```
Comment (prajwal1210) [inline yalis/engine.py]: I don't think this works. done_mask is only defined for batch_size > 1, right now. Need to define it for all batch sizes, right?

**Review by prajwal1210 — APPROVED** [review]
